torch/csrc/distributed/rpc/request_callback_no_python.cpp

```diff
@@ -109,7 +108,16 @@ c10::intrusive_ptr<JitFuture> RequestCallbackNoPython::processMessage(
         },
         c10::getCustomClassType<c10::intrusive_ptr<Message>>());
 
-    return retFuture;
+    auto retFutureWithMessageId = retFuture->then(
+        [id = request.id()](JitFuture& future) {
+          c10::intrusive_ptr<Message> message =
+              future.value().toCustomClass<Message>();
+          message->setId(id);
+          return message;
+        },
+        c10::getCustomClassType<c10::intrusive_ptr<Message>>());
```

Comment:
IIUC, this does some type conversion + map lookup + branching. I wonder if it make sense to keel this `c10::ClassTypePtr` in Message implementation and reuse it in other places?

Reply:
I wouldn't worry about performance here. It's true that we need to do a map lookup and some bookkeeping, but only the first time, as the result is then cached for subsequent uses. See the `static` local variable here https://github.com/pytorch/pytorch/blob/df44f015fe1403f0fea54b46736c07550bb85870/aten/src/ATen/core/ivalue.h#L1161-L1169

Moreover, since that function is templated, the type will be resolved at compile time. Hence a reasonable compiler will inline that code and this callsite will basically just become one memory access (plus a check if the cache has been initialized).